File: garden/seeds/java/app/src/test/java/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/util/RxUtilsTest.java
package com.{{company_name}}.android.{{app_package_name_prefix}}.util;

import com.{{company_name}}.android.{{app_package_name_prefix}}.R;
import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}TestRunner;
import com.{{company_name}}.android.{{app_package_name_prefix}}.activity.HomeActivity;
import com.{{company_name}}.android.{{app_package_name_prefix}}.fragment.BaseFragment;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.shadows.support.v4.SupportFragmentTestUtil;
import org.robolectric.util.ActivityController;

import java.util.concurrent.TimeUnit;

import rx.Observable;
import rx.observers.TestSubscriber;
import rx.schedulers.TestScheduler;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith({{app_class_prefix}}TestRunner.class)
public class RxUtilsTest {

    @Test
    public void testConstructor() {
        new RxUtils(); // For code coverage
    }

    @Test
    public void testBindActivity() {
        final Observable<String> observable = Observable.just("").delay(2, TimeUnit.SECONDS);
        final TestSubscriber<String> subscriber = new TestSubscriber<>();

        ActivityController<HomeActivity> controller = Robolectric.buildActivity(HomeActivity.class);

        controller.create();
        controller.resume();

        RxUtils.bindActivity(controller.get(), observable).subscribe(subscriber);
        assertThat(subscriber.isUnsubscribed()).isFalse();

        controller.pause();
        assertThat(subscriber.isUnsubscribed()).isTrue();
    }

    @Test
    public void testBindFragment() throws InterruptedException {
        final Observable<String> observable = Observable.just("").delay(2, TimeUnit.SECONDS);
        final TestSubscriber<String> subscriber = new TestSubscriber<>();

        DummyFragment frag = new DummyFragment();

        SupportFragmentTestUtil.startFragment(frag);

        frag.onResume();
        RxUtils.bindFragment(frag, observable).subscribe(subscriber);
        assertThat(subscriber.isUnsubscribed()).isFalse();

        frag.onPause();
        assertThat(subscriber.isUnsubscribed()).isTrue();
    }

    static class DummyFragment extends BaseFragment {

        @Override
        protected int getLayoutId() {
            return R.layout.act_home;
        }
    }
}

File: garden/seeds/kotlin/app/src/test/kotlin/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/util/RxUtilsTest.kt
package com.{{company_name}}.android.{{app_package_name_prefix}}.util

import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}TestRunner
import com.{{company_name}}.android.{{app_package_name_prefix}}.R
import com.{{company_name}}.android.{{app_package_name_prefix}}.activity.MainActivity
import com.{{company_name}}.android.{{app_package_name_prefix}}.fragment.BaseFragment
import com.{{company_name}}.android.{{app_package_name_prefix}}.fragment.RxFragment
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.shadows.support.v4.SupportFragmentTestUtil
import rx.Observable
import rx.observers.TestSubscriber
import rx.schedulers.TestScheduler
import java.util.*
import java.util.concurrent.TimeUnit


@RunWith({{app_class_prefix}}TestRunner::class)
public class RxUtilsTest {

    @Test
    public fun testNullFilterWithNullInput() {
        val testSubscriber = TestSubscriber<Any>()

        Observable.just<Any>(null).filterNulls().subscribe(testSubscriber)

        testSubscriber.assertNoValues()
        testSubscriber.assertTerminalEvent()
    }

    @Test
    public fun testNullFilterWithNonNullInput() {
        val EXPECTED_OUTPUT = Object()

        val testSubscriber = TestSubscriber<Any>()
        Observable.just<Any>(EXPECTED_OUTPUT).filterNulls().subscribe(testSubscriber)

        testSubscriber.assertValue(EXPECTED_OUTPUT)
        testSubscriber.assertTerminalEvent()
    }

    @Test
    public fun testFlatMapList() {
        val listInput = listOf(123, 456, 789)

        val testSubscriber = TestSubscriber<Int>()
        Observable.just(listInput).flatMapList().subscribe(testSubscriber)

        testSubscriber.assertValues(123, 456, 789)
    }

    @Test
    public fun testBindActivity() {
        val observable = Observable.just("").delay(2, TimeUnit.SECONDS)
        val subscriber = TestSubscriber<String>()

        val controller = Robolectric.buildActivity(javaClass<MainActivity>())

        controller.create()
        controller.resume()

        observable.bind(controller.get()).subscribe(subscriber)
        assertThat(subscriber.isUnsubscribed()).isFalse()

        controller.pause()
        assertThat(subscriber.isUnsubscribed()).isTrue()
    }

    @Test
    throws(InterruptedException::class)
    public fun testBindFragment() {
        val observable = Observable.just("").delay(2, TimeUnit.SECONDS)
        val subscriber = TestSubscriber<String>()

        val frag = DummyFragment()

        SupportFragmentTestUtil.startFragment(frag)

        frag.onResume()
        observable.bind(frag).subscribe(subscriber)
        assertThat(subscriber.isUnsubscribed()).isFalse()

        frag.onPause()
        assertThat(subscriber.isUnsubscribed()).isTrue()
    }

    private class DummyFragment : BaseFragment() {
        override fun getLayoutId(): Int {
            return 0
        }
    }
}
